[8.0] add auth_password_settings from PR#134

hhgabelgaard opened the pull request:
This is an attempt to make a more "clean" commit history of PR#134
Unfortunately I don't know enough git to extract the relevant history from #134 (git git rebase -s recursive -X theirs origin/8.0 didn't do it for me) - I've "just" attributed the first commit to @dhp-denero - I hope that's acceptable

dhp-denero committed:
auth_password_settings module from https://github.com/OCA/server-tools/pull/134 by dhp-denero

hhgabelgaard committed:
Spelling

hhgabelgaard committed:
Partiel translation to Danish

hhgabelgaard committed:
Changed default min length to 4 to avoid CI errors in auth_supplier
tests and other "test accounts" (e.g. demo/demo)

hbrunn commented on auth_password_settings/README.rst:
```diff
@@ -0,0 +1,22 @@
+Enforce users to set password more complex
```
use OCA's template: https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

hbrunn commented on auth_password_settings/models/res_config.py:
```diff
@@ -0,0 +1,110 @@
+from openerp.tools.translate import _
+
+
+class BaseConfigSettings(osv.osv_memory):
```
use v8 api, this applies to all files

hbrunn commented on auth_password_settings/models/res_users.py:
```diff
@@ -0,0 +1,118 @@
+    def _validate_password(self, cr, uid, password, context=None):
+        icp = self.pool.get('ir.config_parameter')
+        password_rules = []
+        config_data = {
```
call `get_default_auth_password_has_digit` instead of duplicating code

hbrunn commented on auth_password_settings/models/res_users.py:
```diff
@@ -0,0 +1,118 @@
+        return problems
+
+    def write(self, cr, uid, ids, values, context=None):
+        if('password' in values):
```
no brackets

hbrunn commented on auth_password_settings/models/res_users.py:
```diff
@@ -0,0 +1,118 @@
+        if('password' in values):
+            problems = self._validate_password(
+                cr, uid, values['password'], context=context)
+            if(problems):
```
give `_validate_password` a keyword argument indicating it should raise and raise there to have this code centralized

hhgabelgaard committed:
Updated to V8 api

hhgabelgaard committed:
lint on README.rst

hhgabelgaard committed:
pylint

hhgabelgaard committed:
pylint

hhgabelgaard committed:
pylint

hhgabelgaard commented:
How can I get this PR reviewed? Do I miss something? How can I update the labels, i.e. remove "needs fixing" - Or is there still something to fix/improve? 
Thanks!

StefanRijnhart commented:
I see that you have made some changes 29 days ago, but at the time you did not drop a comment saying that you did so. Without making such a comment, the reviewers won't be notified of any changes and are not likely to continue the review.

hhgabelgaard commented:
@StefanRijnhart Thanks for the feedback. I've still have a lot to learn about Github and this process ;-)

bealdav commented on auth_password_settings/README.rst:
```diff
@@ -0,0 +1,83 @@
+{project_repo}/issues/new?body=module:%20
+{module_name}%0Aversion:%20
+{branch}%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+
```
please replace {placeholders} by their real values

bealdav commented on auth_password_settings/README.rst:
```diff
@@ -0,0 +1,83 @@
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+To contribute to this module, please visit https://odoo-community.org.
```
Make sure this readme is update with the model
https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

bealdav commented on auth_password_settings/__init__.py:
```diff
@@ -0,0 +1,21 @@
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
Header template has evolved to a shorter version
https://github.com/OCA/maintainer-tools/blob/master/template/module/exceptions.py#L1-L3

and the header in __init__.py is not mandatory because only import code and chars are only ascii

bealdav commented on auth_password_settings/__openerp__.py:
```diff
@@ -0,0 +1,54 @@
+
+        # Views
+        "views/res_config_view.xml",
+        # Menus
```
remove #Menus and #wizard, etc

bealdav commented on auth_password_settings/models/res_config.py:
```diff
@@ -0,0 +1,93 @@
+                )),
+            'auth_password_has_special_letter': safe_eval(
+                icp.get_param(
+                    'auth_password_settings.auth_password_has_special_letter',
```
What is special character in utf8 context ? Seems you want to say punctuation, isn't it ? Or non ascii ?
Explicit could be better, thanks.

bealdav commented:
Thanks a lot @hhgabelgaard and @dhp-denero for your contribution.

hhgabelgaard committed:
Password change from preference

Issue541

sergio-teruel commented:
@hhgabelgaard please review flake8 errors

hhgabelgaard committed:
Runbot test

hhgabelgaard committed:
Web controllers added as some ValidationErros wasn't shown to the user

hhgabelgaard committed:
Cleanup

hhgabelgaard committed:
flake

Conflicts:
	auth_password_settings/controllers/main.py
	auth_password_settings/models/res_users.py

hhgabelgaard committed:
flake

hhgabelgaard committed:
flake

hhgabelgaard committed:
flake

hhgabelgaard commented:
@bealdav and @sergio-teruel : Thanks for the review
I've updated the module too also support password reset from the login page - and fixed flake errors etc

dreispt commented:
There is an alternative implementation at #350 
Please discuss there the path to be followed.

dreispt commented on auth_password_settings/README.rst:
```diff
@@ -0,0 +1,80 @@
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/{repo_id}/{branch}
```
Replace the placeholders: repo_id is 149 and branch is 8.0

dreispt commented on auth_password_settings/__openerp__.py:
```diff
@@ -0,0 +1,29 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    'name': 'User Password Settings',
+    'version': '1.0',
```
8.0.1.0.0

dreispt commented on auth_password_settings/__openerp__.py:
```diff
@@ -0,0 +1,29 @@
+{
+    'name': 'User Password Settings',
+    'version': '1.0',
+    'category': 'base',
```
I'm pretty sure this is not correct.

dreispt commented on auth_password_settings/__openerp__.py:
```diff
@@ -0,0 +1,29 @@
+        "views/res_config_view.xml",
+        ],
+    'installable': True,
+    'application': True,
```
This is a  feature, not an app. Remove please.

dreispt commented on auth_password_settings/data/auth_password_settings_data.xml:
```diff
@@ -0,0 +1,22 @@
+        <!--SETTINGS-->
+        <record id="default_auth_password_min_character" model="ir.config_parameter">
+            <field name="key">auth_password_settings.auth_password_min_character</field>
+            <field name="value">4</field>
```
Quite conservative default. I would have this to 10 or 12.

hhgabelgaard replied:
It was 6 (like the default), but then Ttravis failed on other modules in this repos adding test accounts like demo/demo etc

dreispt commented on auth_password_settings/models/res_config.py:
```diff
@@ -0,0 +1,93 @@
+            'auth_password_min_character': safe_eval(
+                icp.get_param(
+                    'auth_password_settings.auth_password_min_character',
+                    '6'
```
Isn't it more reasonable to use the neutral 0 if no setting is found?

dreispt commented on auth_password_settings/models/res_config.py:
```diff
@@ -0,0 +1,93 @@
+        self.ensure_one()
+        if self.auth_password_min_character < 5:
+            raise ValidationError(
+                _('Password Length should not be less then 5.')
```
Why not? It could be zero!
IMO this validation should go away.

dreispt commented on auth_password_settings/views/res_config_view.xml:
```diff
@@ -0,0 +1,42 @@
+                        <div>
+                            <p>
+                                <label string="This section allows you to determine password complexity (the effectiveness of a password in resisting guessing and brute-force attacks). To do that," /><br/>
+                                <label string="* use the Minimum password length to determine how long the password should be to be considered strong, By default, the password should be at least 6 characters long."/><br/>
```
I'm not sure if we should keep the "By default, the password should be at least 6 characters long."

dreispt commented:
Nice work. 
Functionally, the minimum 6 digit limit seems arbitrary, and IMO there is no point in enforcing it.
Technically, a few nitpicks to address.
Style-wise, I personally dislike the Java-like closing braces and brackets, but I can't pick on that.

hhgabelgaard committed:
Merge branch '8.0-auth_password_settings-hhg-541' of https://github.com/steingabelgaard/server-tools.git into 8.0-auth_password_settings-hhg

hhgabelgaard committed:
Removed the default length of 6 characters and related changes pr
@dreispt suggestions

hhgabelgaard committed:
Pylint: Version number

hhgabelgaard committed:
flake

hhgabelgaard committed:
flake

hhgabelgaard commented:
@dreispt Thank you for the review - I hope I got most of it right ;-)

StefanRijnhart commented on auth_password_settings/README.rst:
```diff
@@ -0,0 +1,76 @@
+help us smashing it by providing a detailed and welcomed `feedback
+<https://github.com/OCA/server-tools/issues/new?body=module:%20auth_password_settings%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+
+Credits
```
Can you please add to this section where you found the icon?

hhgabelgaard replied:
The icon is from @dhp-denero original PR #134 - I don't know where he got it
Probaly better to replace it with the OCA icon?

StefanRijnhart commented on auth_password_settings/controllers/main.py:
```diff
@@ -0,0 +1,97 @@
+
+class SessionEx(Session):
+
+    @http.route('/web/session/change_password', type='json', auth="user")
```
You are breaking the inheritance chain here. It seems that it might be possible to perform checks in an override before or after calling super. Did you look into that?

StefanRijnhart commented on auth_password_settings/controllers/main.py:
```diff
@@ -0,0 +1,97 @@
+
+    @http.route('/web/reset_password', type='http',
+                auth='public', website=True)
+    def web_auth_reset_password(self, *args, **kw):
```
While you call super here, you do copy more code from auth_signup than seems necessary. How about calling super immediately, and then validate the password if you have a password in qcontext and there is no error defined?

dreispt commented:
@hhgabelgaard Hi there, a new competing PR was proposed for similar features: please see #531 
You're welcome to comment there.

hhgabelgaard commented:
#531 seems to be a better implementation than this code .
That being said, we have been running this code in production for 8 months now with 1000+ users without problems.

dreispt commented:
As discussed, this was overseeded by #531.
It targets 9.0 so contributions to backport it to 8.0 are welcome.
